File: LeetCode160.cpp
#include <vector>
#include<algorithm>
#include<iostream>
using namespace std;

typedef struct ListNode{
    int val;
    struct ListNode *next;
}ListNode, *LinkList;

class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        if (headA == nullptr || headB == nullptr) {
            return nullptr;
        }
        ListNode *pA = headA, *pB = headB;
        while (pA != pB) {
            pA = pA == nullptr ? headB : pA->next;
            pB = pB == nullptr ? headA : pB->next;
        }
        return pA;
    }
    LinkList CreateList2(LinkList &head1, LinkList &head2, vector<int> nums1, vector<int> nums2, vector<int> pub){
        int n = pub.size();
        LinkList L=(LinkList)malloc(sizeof(ListNode));
        ListNode *s,*r=L;
        // r->val = nums1[0];
        // i = 1;
        int i = 0;
        while(i < n){
            s=(ListNode*)malloc(sizeof(ListNode));
            s->val=pub[i++];
            r->next=s;
            r=s;
        }
        // 也可以r->next = nullptr;
        s->next = nullptr;
        //-------------------------------//
        n = nums1.size();
        r = head1;
        i = 0;
        while(i < n){
            s=(ListNode*)malloc(sizeof(ListNode));
            s->val=nums1[i++];
            r->next=s;
            r=s;
        }
        s->next = L->next;

        //-------------------------------//
        n = nums2.size();
        r = head2;
        i = 0;
        while(i < n){
            s=(ListNode*)malloc(sizeof(ListNode));
            s->val=nums2[i++];
            r->next=s;
            r=s;
        }
        s->next = L->next;

        return nullptr;
    }
};

int main(){
    vector<int> nums1, nums2, pub;
    int n = 0, t = 0;
    cout << "number of num1: ";
    cin >> n;
    for (int i = 1; i <= n; i++){
        cin >> t;
        nums1.push_back(t);
    }
    cout << "number of num2: ";
    cin >> n;
    for (int i = 1; i <= n; i++){
        cin >> t;
        nums2.push_back(t);
    }
    cout << "number of public: ";
    cin >> n;
    for (int i = 1; i <= n; i++){
        cin >> t;
        pub.push_back(t);
    }
    ListNode *head2 = (LinkList)malloc(sizeof(ListNode));
    ListNode* head1 = (LinkList)malloc(sizeof(ListNode));
    Solution sol;
    sol.CreateList2(head1, head2, nums1, nums2, pub);
    sol.getIntersectionNode(head1, head2);
    return 0;
}
// number of nums1: 4
// 1 3 5 7
// number of nums1: 4
// 2 4 6 8
// number of public: 1
// 1

File: LeetCode160.java
import java.util.Scanner;

public class LeetCode160 {
    public static void main(String[] args) {
        // 一维整数数组
        LeetCode160 solution = new LeetCode160();
        Scanner scanner = new Scanner(System.in);

        int[] numbers1, numbers2;
        int n = 0, m = 0;
        System.out.print("number of numbers1: ");
        m = scanner.nextInt();
        System.out.print("number of numbers2: ");
        n = scanner.nextInt();
        numbers1 = new int[m];
        for (int i = 1; i <= m; i++){
            numbers1[i-1] = scanner.nextInt();
        }
        numbers2 = new int[n];
        for (int i = 1; i <= n; i++){
            numbers2[i-1] = scanner.nextInt();
        }

        System.out.print("number of public: ");
        n = scanner.nextInt();
        int[] publicNum = new int[n];
        for (int i = 1; i <= n; i++){
            publicNum[i-1] = scanner.nextInt();
        }
        //非递归解决
        ListNode L1 = new ListNode();
        ListNode L2 = new ListNode();
        //尾插；带头结点
        solution.CreateList2(L1, numbers1, L2, numbers2, publicNum);
        solution.getIntersectionNode(L1.next, L2.next);
        scanner.close();
    }

    ListNode CreateList2(ListNode L1, int[] nums1, ListNode L2, int[] nums2, int[] pub){
        int n = pub.length;
        ListNode s = new ListNode();
        ListNode p = new ListNode();
        ListNode r = p;
        int i = 0;
        while(i < n){
            s = new ListNode();
            s.val = nums1[i++];
            r.next = s;
            r=s;
        }
        s.next = null;

        n = nums1.length;
        s = new ListNode();
        r = L1;
        i = 0;
        while(i < n){
            s = new ListNode();
            s.val = nums1[i++];
            r.next = s;
            r=s;
        }
        s.next = p.next;

        n = nums2.length;
        s = new ListNode();
        r = L2;
        i = 0;
        while(i < n){
            s = new ListNode();
            s.val = nums2[i++];
            r.next = s;
            r=s;
        }
        s.next = p.next;

        return null;
    }
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if (headA == null || headB == null) {
            return null;
        }
        ListNode pA = headA, pB = headB;
        while (pA != pB) {
            pA = pA == null ? headB : pA.next;
            pB = pB == null ? headA : pB.next;
        }
        return pA;
    }
}

/*给你两个单链表的头节点 headA 和 headB ，请你找出并返回两个单链表相交的起始节点。如果两个链表不存在相交节点，返回 null 。*/
// number of nums1: 4
// 1 3 5 7
// number of nums2: 4
// 2 4 6 8
// number of public: 1
// 1
// number of nums1: 4
// 1 3 5 7
// number of nums2: 4
// 2 4 6 8
// number of public: 1
// 1